### ISSUE TITLE
"abusing renderOrder" potential performance issues?

### ISSUE THREAD

**Issue by slosumo** [KilledByAPixel/LittleJS#22]
TLDR: Is it safe to use *any* amount of different render orders, or is it recommended for some sort of reason to stay under a certain amount?

I'm wondering about something, and if it might cause potential performance issues -> I think not, as the way renderOrder works is just a sort, but just wanted to double check in case I'm missing some intricacy of how it all works under the hood.

Essentially, if I have objects with the same renderOrder, they will be rendered in order of their creation, which can cause situation like this: (top tree being rendered on top):
![image](https://user-images.githubusercontent.com/10269798/145202425-27c29605-480d-49b4-a5a8-db9dccaa101b.png)

Now this is easily solvable, if I give every tree a different render order, based on their position (i.e. rendering goes from higher pos.y down to the lowest, and from lowest pos.x to the highest), but i.e. if I had a map of 50x50 then that's potentially 2500 different render orders. I suppose that shouldn't be an issue, as the render order is just a matter of sorting the integers?

Essentially I want to render objects on different layers (but without using tileLayers as they're often animated etc), and I could keep diff objects layered properly, and avoid these render issues like so:
Say on a 10x10 map, I have undergrowth & trees layer, I could then say that all:
```
Undergrowth have a renderOrder of  0 + (100 - (pos.x + levelSize.x * pos.y)) =  (0 - 99)
Trees have a base a renderOrder of 100 + (100 - (pos.x + levelSize.x * pos.y)) (100 - 199)
```

And this should work with basically any map size and any amount of different layers (besides obviously hardware limit of how many can be rendered at the same time)

**Comment by KilledByAPixel** [comment]
You got it all figured out, that should work great. 

I intended that renderOrder could be used to make this kind of game. I did something similar in my game Bounce Back where I basically just used the Y position as the render order.

It's just a simple sort, I really tried to keep everything simple with this engine. with your game, you could probably get away with not sorting it every frame because objects dont move around, but it really isn't that much overhead.

Keep up the good work, I'm happy to answer any more questions that come up.

**Comment by slosumo** [comment]
Cool thanks for confirming - it works great indeed and no performance hits indeed!
I've also moved away from using most tileLayers where animating stuff often, as it simplifies so much, as long as we keep some level size constraints in mind (i.e. max 10k "objects" in render view).

**Btw a tip for documentation, maybe add a performance tips section or something:**
Your stress test initially kinda moved me away from using engine objects as it could handle only 1-2k. But I realized that if you simply disable collission/gravity etc. for them (or just not call super.update()), it easily handles 50-70% amount of the sprites only test (instead of like 5-10%) - i.e. on FF for me 25k at 60fps.

25k is a ton for even top down games that are heavily animated, and using classes extending engineObject and the internal rendering etc. makes it just super convenient to use - of course one could write completely custom render logic like you do for the sprites test if additional performance is needed, but honestly it looks like even I won't be needing that with bigger maps than most use-cases :) It really simplified it a lot for me once I figured this out, and can now easily keep things more organized and decoupled and not worry about performance with 10-20k things being rendered :)

**Comment by KilledByAPixel** [comment]
Thats great.

I think a performance tips in the documentation would be great. I really need to write a page or two about how to get the best out of the engine. All the stuff you are talking about is helping me learn what to include.
